File: scripts/optimize_images.py
#!/usr/bin/env python3
"""Optimize Locia Farms website images.

- Creates WebP versions next to original files (same basename + .webp)
- Resizes to max width (default 2000px) to prevent huge payloads
- Preserves originals (no destructive edits)

Usage:
  python3 scripts/optimize_images.py --root images --max-width 2000 --quality 80
"""

from __future__ import annotations

import argparse
from pathlib import Path

from PIL import Image


def iter_images(root: Path):
    exts = {".jpg", ".jpeg", ".png", ".JPG", ".JPEG", ".PNG"}
    for p in root.rglob("*"):
        if p.is_file() and p.suffix in exts:
            yield p


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--root", type=Path, default=Path("images"))
    ap.add_argument("--max-width", type=int, default=2000)
    ap.add_argument("--quality", type=int, default=80)
    ap.add_argument("--force", action="store_true", help="overwrite existing .webp")
    args = ap.parse_args()

    root = args.root
    if not root.exists():
        raise SystemExit(f"Root not found: {root}")

    count = 0
    saved_kb = 0
    for src in iter_images(root):
        # Skip already-optimized derivatives
        if src.suffix.lower() == ".webp":
            continue

        dst = src.with_suffix(".webp")
        if dst.exists() and not args.force:
            continue

        try:
            im = Image.open(src)
        except Exception:
            continue

        # Convert to RGB for webp
        if im.mode in ("RGBA", "LA"):
            # WebP supports alpha; keep if present
            pass
        elif im.mode != "RGB":
            im = im.convert("RGB")

        w, h = im.size
        if w > args.max_width:
            new_h = int(h * (args.max_width / w))
            im = im.resize((args.max_width, new_h), Image.Resampling.LANCZOS)

        # Save webp
        dst.parent.mkdir(parents=True, exist_ok=True)
        im.save(dst, format="WEBP", quality=args.quality, method=6)

        # Rough savings
        try:
            saved_kb += max(0, src.stat().st_size - dst.stat().st_size) // 1024
        except Exception:
            pass

        count += 1

    print(f"Generated {count} .webp files; approx saved {saved_kb} KB")


if __name__ == "__main__":
    main()

File: scripts/optimize-images.mjs
import fs from 'node:fs/promises';
import path from 'node:path';
import sharp from 'sharp';

const targets = [
  {
    src: 'images/splash-home-01.jpg',
    widths: [800, 1200, 1600],
    formats: ['webp', 'jpeg'],
  },
  {
    src: 'images/splash-home-goat-01.jpeg',
    widths: [800, 1200],
    formats: ['webp'],
  },
  {
    src: 'images/splash-home-chickens-01.jpeg',
    widths: [800, 1200],
    formats: ['webp'],
  },
  {
    src: 'images/splash-home-02.jpeg',
    widths: [800, 1200],
    formats: ['webp'],
  },
];

function outPath(src, width, fmt) {
  const ext = path.extname(src);
  const base = src.slice(0, -ext.length);
  return `${base}.${width}w.${fmt}`;
}

async function exists(p) {
  try { await fs.stat(p); return true; } catch { return false; }
}

let changed = 0;
for (const t of targets) {
  if (!(await exists(t.src))) {
    console.error(`Missing: ${t.src}`);
    continue;
  }

  const img = sharp(t.src);
  const meta = await img.metadata();

  for (const w of t.widths) {
    if (meta.width && meta.width < w) continue;

    for (const fmt of t.formats) {
      const out = outPath(t.src, w, fmt);
      if (await exists(out)) continue;

      let pipeline = sharp(t.src).resize({ width: w, withoutEnlargement: true });
      if (fmt === 'webp') pipeline = pipeline.webp({ quality: 82 });
      if (fmt === 'jpeg') pipeline = pipeline.jpeg({ quality: 82, mozjpeg: true });

      await pipeline.toFile(out);
      changed++;
      console.log(`Wrote ${out}`);
    }
  }
}

console.log(`Done. New files: ${changed}`);
